src/main/java/graphql/validation/RulesVisitor.java

```diff
@@ -88,139 +79,108 @@ public void enter(Node node, List<Node> ancestors) {
-        for (AbstractRule rule : rules) {
-            rule.checkTypeName(node);
-        }
+        rules.forEach(r -> r.checkTypeName(node));
     }
 
-
-    private void checkVariableDefinition(VariableDefinition variableDefinition, List<AbstractRule> rules) {
-        for (AbstractRule rule : rules) {
-            rule.checkVariableDefinition(variableDefinition);
-        }
+    private void checkVariableDefinition(VariableDefinition node, List<AbstractRule> rules) {
+        rules.forEach(r -> r.checkVariableDefinition(node));
     }
 
-    private void checkField(Field field, List<AbstractRule> rules) {
-        for (AbstractRule rule : rules) {
-            rule.checkField(field);
-        }
+    private void checkField(Field node, List<AbstractRule> rules) {
+        rules.forEach(r -> r.checkField(node));
     }
 
-    private void checkInlineFragment(InlineFragment inlineFragment, List<AbstractRule> rules) {
-        for (AbstractRule rule : rules) {
-            rule.checkInlineFragment(inlineFragment);
-        }
+    private void checkInlineFragment(InlineFragment node, List<AbstractRule> rules) {
+        rules.forEach(r -> r.checkInlineFragment(node));
     }
 
-    private void checkDirective(Directive directive, List<Node> ancestors, List<AbstractRule> rules) {
-        for (AbstractRule rule : rules) {
-            rule.checkDirective(directive, ancestors);
-        }
+    private void checkDirective(Directive node, List<Node> ancestors, List<AbstractRule> rules) {
+        rules.forEach(r -> r.checkDirective(node, ancestors));
     }
 
-    private void checkFragmentSpread(FragmentSpread fragmentSpread, List<AbstractRule> rules, List<Node> ancestors) {
-        for (AbstractRule rule : rules) {
-            rule.checkFragmentSpread(fragmentSpread);
-        }
-        List<AbstractRule> rulesVisitingFragmentSpreads = getRulesVisitingFragmentSpreads(rules);
+    private void checkFragmentSpread(FragmentSpread node, List<AbstractRule> rules, List<Node> ancestors) {
+        rules.forEach(r -> r.checkFragmentSpread(node));
+
+        ImmutableList<AbstractRule> rulesVisitingFragmentSpreads = filterRulesVisitingFragmentSpreads(rules, true);
         if (rulesVisitingFragmentSpreads.size() > 0) {
-            FragmentDefinition fragment = validationContext.getFragment(fragmentSpread.getName());
-            if (fragment != null && !ancestors.contains(fragment)) {
-                new LanguageTraversal(ancestors).traverse(fragment, new RulesVisitor(validationContext, rulesVisitingFragmentSpreads, true));
+            FragmentDefinition fragment = validationContext.getFragment(node.getName());
+            if (fragment != null && !ancestors.contains(fragment) && !visitedFragmentSpreads.contains(node.getName())) {
+                // Manually traverse into the FragmentDefinition
+                visitedFragmentSpreads.add(node.getName());
+                rulesStack.push(rulesVisitingFragmentSpreads);
```

Comment:
just as an example why I think the current code is too complicated: this is always the same set of 3 rules, we don't need to maintain an extra stack for that. 

Btw: the old code before your PR was not better, nothing todo with your PR directly :-)